packages/snaps-utils/src/snaps.ts

```diff
@@ -195,6 +207,65 @@ export function validateSnapShasum(
   }
 }
 
+export const LOCALHOST_HOSTNAMES = ['localhost', '127.0.0.1', '::1'] as const;
+
+type UriOptions<T extends string> = {
+  protocol?: Struct<T>;
+  hash?: Struct<T>;
+  port?: Struct<T>;
+  hostname?: Struct<T>;
+  pathname?: Struct<T>;
+  search?: Struct<T>;
+};
+export const uri = (opts: UriOptions<any> = {}) =>
+  define<string | URL>('uri', (value) => {
+    try {
+      const url = new URL(value as any);
+
+      const UrlStruct = type(opts);
+      assertSuperstruct(url, UrlStruct);
+      return true;
+    } catch {
+      return `Expected URL, got "${(value as any).toString()}"`;
+    }
+  });
+
+const LocalSnapIdSubUrlStruct = uri({
+  protocol: enums(['http:', 'https:']),
+  hostname: enums(LOCALHOST_HOSTNAMES),
+  hash: empty(string()),
```

Comment:
Do we need this? As far as I know it's never part of an HTTP request.

Reply:
Yes. Otherwise you could have something like `local:localhost:8080/#asd` as a valid snap ID. It would be different snap ID from `local:localhost:8080/`